esphome/components/http_request/http_request.cpp

```diff
@@ -115,10 +115,10 @@ void HttpRequestComponent::close() {
 }
 
 const char *HttpRequestComponent::get_string() {
-  // The static variable is here because HTTPClient::getString() returns a String on ESP32, and we need something to
-  // to keep a buffer alive.
-  static std::string str;
-  str = this->client_.getString().c_str();
+  // The static variable is here because HTTPClient::getString() returns a String on ESP32,
+  // and we need something to keep a buffer alive.
+  static String str;
+  str = this->client_.getString();
```

Comment:
How about something like that?
```suggestion
#if defined (ESP32)
  // The static variable is here because HTTPClient::getString() returns a String on ESP32,
  // and we need something to keep a buffer alive.
  static String str;
#else
  // However on ESP8266, HTTPClient::getString() returns a String& to a member variable.
  // Leaving this the default so that any new platform either doesn't copy, or encounters a compilation error.
  String&
#endif
  str = this->client_.getString();
```

Reply:
or maybe use `auto` in the `#else` clause, to allow additional return types like `std::string&`?
```suggestion
#if defined (ESP32)
  // The static variable is here because HTTPClient::getString() returns a String on ESP32,
  // and we need something to keep a buffer alive.
  static String str;
#else
  // However on ESP8266, HTTPClient::getString() returns a String& to a member variable.
  // Leaving this the default so that any new platform either doesn't copy, or encounters a compilation error.
  auto
#endif
  str = this->client_.getString();
```

Reply:
We could, though the proper fix for this mess would be to change this whole method to return a `std::string` and transfer ownership to the caller.

Reply:
Probably `String` and not `std::string`- to avoid copy.
I agree, that was my initial suggestion in esphome/issues#2903, but it breaks the API. Is that allowed?

Since the current comment of `HTTPClient::getString()` in both ESP32 and ESP8266 code says "`(may need lot of ram or trigger out of memory!)`", I'd hate to leave this code in a state that temporarily consumes twice the required amount..

Reply:
We try to avoid the Arduino String type in esphome.
Also if we manage to write an idf http component returning an Arduino String (reference) isn't optional.
So I agree it's not the optimal solution but I don't see better alternatives.

Reply:
Yeah, we don't use Arduino types in APIs in ESPHome. 

I think the copy could be avoided by implementing a `Stream` subclass that writes to a `std::string`, and using the `writeToStream` function, but honestly that effort is probably better invested in switching away from the Arduino HTTP libraries. They're not exactly high-quality, and we need to eventually do that for ESP-IDF anyway.

Reply:
What's the current state of this? I'm fine with the PR as-is or either suggestion, what do you prefer? Let's get this one done :)

Reply:
Sorry, I was under the impression that we're waiting for your decision between the options.

Anyway, I've been experimenting a little more (and spotted a mistake in one of my suggestions - edited above already), pondering what's the right default (for a theoretical future platform which shares a compatible API but is not ESP32 nor ESP8266).

if a new platform/API returns by value: (like ESP32, possibly a different type though)
* `auto&` will cause a compilation error
* `auto` or `auto&&' will produce wrong behavior, and should output a compilation warning.
* `static String` will work correctly, but will use copy if it returns std::string

if a new platform/API returns a reference: (like ESP8266)
* `auto&` or `auto&&` works great and avoids copy
* `auto` will produce wrong behavior, and should output a compilation warning.
* `static String` will work correctly, but use copy

I also considered `static std::remove_reference_t<decltype(this->client_.getString())>` (which will use 'move' both with `String` and `std::string`, but the syntax looks too scary and I don't like that it repeats the call statement.

So it narrows down to deciding what is preferred on future incompatibility - compilation error vs. silent hidden copy. (former means `auto&`, latter means `static String`)
Initially I preferred the compilation error aproach, but thinking again at how much effort is required when adopting the code to a new platform, I thought maybe it's better to make the maintainer's life easier and potentially optimize later on.

What is the preferred spirit in this project? Tell me and I'll update the PR accordingly and test it (to be optimal for both ESP32 and ESP8266, and what you choose as default)

Reply:
Sorry, I was under the impression that we're waiting for your decision between the options.

Anyway, I've been experimenting a little more (and spotted a mistake in one of my suggestions - edited above already), pondering what's the right default (for a theoretical future platform which shares a compatible API but is not ESP32 nor ESP8266).

I think I'll go with the `auto&` option, I'll further test and prepare it soon.

I also considered `static std::remove_reference_t<decltype(this->client_.getString())>` (which will work correctly both with `String` and `std::string`, but (A) the syntax looks too scary, (B) I don't like that it repeats the call statement, and (C) it uses copy on non-ESP8266 platforms that return a reference.

Reply:
It might very well have been blocked on me, I kinda lost track of this one :)

> I think I'll go with the `auto&` option for ESP8266 and default, I'll further test and prepare it soon.

Great!